### DISCUSSION TITLE
Can an addon contribute other addons?

### DISCUSSION THREAD

**Discussion by jd-carroll** [storybookjs/storybook#26833] [Help] [answered]
### Summary

We are working with a mono-repo with storybook configurations in most of the modules.  Right now our `main.ts` and `preview.tsx` look like:

```typescript
// main.ts
import { buildMainConfig } from '@project/client-storybook-utils';

export default buildMainConfig();
```
```typescript
// preview.tsx
import { PreviewConfig } from "@project/client-storybook-utils/lib/PreviewConfig";

export default PreviewConfig;
```

It works, but it is not statically analyzable so we end up with these giant `CSF Parse error`'s. It certainly works if we put the config in each module, but becomes a pain to maintain.

**I would like to use a custom addon, but I can't see to find any documentation that describes how an addon can contribute other addons. Is this possible? If so, could someone point me to an example?**

If this is not possible, how do people manage "shared" configurations across a monorepo / for internal projects?

**Comment by valentinpalkovic** [comment, marked as answer]
Hi @jd-carroll 

@storybook/addon-essentials does exactly that. An addon can export an `addons` preset field, which will be taken into account by Storybook to load further addons:

https://github.com/storybookjs/storybook/blob/next/code/addons/essentials/src/index.ts#L67

**Reply by jd-carroll** [reply]
Amazing, thank you! Will start playing with that, but I think there is a hole in the documentation. Please don't get me wrong, I am incredibly grateful for all the work that has / is going into Storybook, but... There doesn't seem to be any kind of API documentation. The documentation just focuses on "this is how you do things..."

### Pieces I think need addressing:

##### Is it an add-on or something else?
What is the difference between an add-on and a preset / preview / manager contribution? Conceptually I think everything is an add-on and an add-on can contribute a preset / preview / manager (but maybe this is actually wrong)

##### What are the entry points?
How does storybook load all of these pieces? I think there needs to be something like the [node module loader documentation](https://nodejs.org/docs/latest/api/modules.html#all-together) to just keep it explicit.  Additionally there should be clear API interfaces where the parameter definitions (if any) are also explicit.

##### How is the local project different?
This one maybe a little unfair, because I do think Storybook does a good job sharing the amount of power you have contributing one-off parts in your local setup. However, there doesn't seem to be a unifying concept across all of them. Its almost (not quite, but almost) like there are add-ons / presets / previews / managers  which are only available through third-party modules. And then separately there are decorators / loaders / play which are local project concepts.

### How can these be addressed:
To be clear, I don't think there needs to be any API changes. I think the unifying pieces / concepts are clear to the core Storybook team, but those concepts aren't completely captured in the documentation.

Personally, I think everything needs to be an add-on. Then it can be distilled to:
1) This is how you write a shareable add-on
2) This is how you write a "local add-on" in your project

I get that the semantics of writing a "local add-on" might be a little off and that the way it works is technically probably a little different. But from a conceptual standpoint, it would make the documentation much more clear.

All that to say: **Thank you for the link!**
